src/Services/PaymentReferenceCodeGenerator/UniquePaymentReferenceCodeGenerator.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\PaymentContext\Services\PaymentReferenceCodeGenerator;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\ORM\EntityManager;
+use WMDE\Fundraising\PaymentContext\Domain\CharacterIndexGenerator;
+use WMDE\Fundraising\PaymentContext\Domain\Model\PaymentReferenceCode;
+use WMDE\Fundraising\PaymentContext\Domain\PaymentReferenceCodeGenerator;
+
+class UniquePaymentReferenceCodeGenerator implements PaymentReferenceCodeGenerator {
+
+	private PaymentReferenceCodeGenerator $randomPaymentReferenceGenerator;
+	private Connection $connection;
+
+	public function __construct( CharacterIndexGenerator $characterIndexGenerator, EntityManager $entityManager ) {
+		$this->randomPaymentReferenceGenerator = new RandomPaymentReferenceCodeGenerator( $characterIndexGenerator );
```

Comment:
IMHO this class should be a wrapper/decorator. You can pass in the new `PaymentReferenceCode` interface instead of creating it in the constructor. It'll make testing more expressive (by using explicit sequential payment reference codes that provide uniqueness) and this class more reusable.